src/hotspot/share/prims/whitebox.cpp

```diff
@@ -778,26 +778,31 @@ WB_ENTRY(jboolean, WB_IsFrameDeoptimized(JNIEnv* env, jobject o, jint depth))
 WB_END
 
 WB_ENTRY(void, WB_DeoptimizeAll(JNIEnv* env, jobject o))
-  CodeCache::mark_all_nmethods_for_deoptimization();
-  Deoptimization::deoptimize_all_marked();
+  DeoptimizationScope deopt_scope;
+  CodeCache::mark_all_nmethods_for_deoptimization(&deopt_scope);
+  deopt_scope.deoptimize_marked();
 WB_END
 
 WB_ENTRY(jint, WB_DeoptimizeMethod(JNIEnv* env, jobject o, jobject method, jboolean is_osr))
   jmethodID jmid = reflected_method_to_jmid(thread, env, method);
   int result = 0;
   CHECK_JNI_EXCEPTION_(env, result);
-  MutexLocker mu(Compile_lock);
-  methodHandle mh(THREAD, Method::checked_resolve_jmethod_id(jmid));
-  if (is_osr) {
-    result += mh->mark_osr_nmethods();
-  } else if (mh->code() != nullptr) {
-    mh->code()->mark_for_deoptimization();
-    ++result;
-  }
-  result += CodeCache::mark_for_deoptimization(mh());
-  if (result > 0) {
-    Deoptimization::deoptimize_all_marked();
+
+  DeoptimizationScope deopt_scope;
+  {
+    MutexLocker mu(Compile_lock);
+    methodHandle mh(THREAD, Method::checked_resolve_jmethod_id(jmid));
+    if (is_osr) {
+      result += mh->method_holder()->mark_osr_nmethods(&deopt_scope, mh());
+    } else if (mh->code() != nullptr) {
+      deopt_scope.mark(mh->code());
```

Comment:
I was working on a patch based on top of this one and had a crash here.

The second call to `mh->code()` returned a `nullptr`. It looks racy to me and seems like the only thing protecting `CompileMethod::code()` is the `CompiledMethod_lock` which is not held here. 

Regardless `mh->code()` should probably only be loaded once. (Or at least not re-loaded after the nullptr check)

Reply:
Fixed, thanks!